### PR TITLE
WIP: usb: usb_dc_stm: Add support for STM32Cube HAL_PCD USB driver

### PR DESCRIPTION
**Pull request by ydamigos**
This is a quick and dirty driver for the USB device controller on STM32F4xx devices using the STM32 Cube HAL_PCD framework developed by Christer Weinigel and Daniel Thompson.

This PR just rebased it against latest arm branch. The driver was tested on 96b_carbon using the  samples/subsys/usb/cdc_acm sample.

The original PR can be found on gerrit: https://gerrit.zephyrproject.org/r/#/c/10176/

### CODE REVIEW

File: drivers/usb/device/usb_dc_stm.c [hunk 1]
```diff
@@ -0,0 +1,644 @@
...
+		return -EIO;
+	}
+
+	usb_dc_stm_state.out_ep_state[EP0_IDX].ep_mps = EP0_MPS;
+	usb_dc_stm_state.out_ep_state[EP0_IDX].ep_type = EP_TYPE_CTRL;
+	usb_dc_stm_state.in_ep_state[EP0_IDX].ep_mps = EP0_MPS;
+	usb_dc_stm_state.in_ep_state[EP0_IDX].ep_type = EP_TYPE_CTRL;
+
+	HAL_PCDEx_SetRxFiFo(&usb_dc_stm_state.pcd, FIFO_SIZE_IN_WORDS);
+	for (i = 0; i < CONFIG_USB_DC_STM_EP_NUM; i++) {
```
Comment (jfischer-phytec-iot) [inline drivers/usb/device/usb_dc_stm.c]: Here again, what gives CONFIG_USB_DC_STM_EP_NUM ?

File: arch/arm/soc/st_stm32/stm32f4/Kconfig.defconfig.stm32f401xe [hunk 2]
```diff
@@ -15,4 +15,12 @@ config NUM_IRQS
 	int
 	default 85
 
+config USB_DC_STM_EP_NUM
+	int
+	default 4
+
+config USB_DC_STM_PACKET_RAM_SIZE
+	int
+	default 1280
```
[the remark this answers is not in the document]
Comment (ydamigos) [reply arch/arm/soc/st_stm32/stm32f4/Kconfig.defconfig.stm32f401xe]: It depends on the SoC. For example, STM32F303VC has 512 bytes packet buffer RAM and 8 endpoints.

**Comment by ydamigos** [comment]
@jfischer-phytec-iot As I am not the original author of the driver, I will try to address your comment while I am studying and understanding the driver.

File: drivers/usb/device/usb_dc_stm.c [hunk 3]
```diff
@@ -0,0 +1,644 @@
...
+#include <usb/usb_device.h>
+#include <clock_control/stm32_clock_control.h>
+
+#define SYS_LOG_LEVEL CONFIG_SYS_LOG_USB_DRIVER_LEVEL
+#include <logging/sys_log.h>
+
+/* Miscellaneous values used in the driver */
+
+/* The maximum packet size for endpoint 0 */
+#define EP0_MPS USB_OTG_MAX_EP0_SIZE
```
[the remark this answers is not in the document]
Comment (ydamigos) [reply drivers/usb/device/usb_dc_stm.c]: I try to make the driver generic for all the STM32 families. USB_OTG_MAX_EP0_SIZE is not available for SOC families with USB device controller.

File: drivers/usb/device/usb_dc_stm.c [hunk 4]
```diff
@@ -0,0 +1,644 @@
...
+
+/* Size of a USB SETUP packet */
+#define SETUP_SIZE 8
+
+/* Divide the packet memory evenly between the RX/TX FIFOs.
+ *
+ * E.g. for the STM32F405 with 1280 bytes of RAM, 1 RX FIFO and 4 TX
+ * FIFOs this will allocate 1280 / 5 = 256 bytes for each FIFO.
+ */
+#define FIFO_SIZE_IN_WORDS ((CONFIG_USB_DC_STM_PACKET_RAM_SIZE / 4) / (1 + CONFIG_USB_DC_STM_EP_NUM))
```
[the remark this answers is not in the document]
Comment (ydamigos) [reply drivers/usb/device/usb_dc_stm.c]: I updated this. The CONFIG_USB_DC_STM_EP_NUM is the number of bidirectional endpoints.

File: drivers/usb/device/usb_dc_stm.c [hunk 5]
```diff
@@ -0,0 +1,644 @@
...
+	return 0;
+}
+
+static int usb_dc_stm_init(void)
+{
+	HAL_StatusTypeDef status;
+	unsigned int i;
+
+	usb_dc_stm_state.pcd.Instance = USB_OTG_FS;
+	usb_dc_stm_state.pcd.Init.dev_endpoints = 3;
```
[the remark this answers is not in the document]
Comment (ydamigos) [reply drivers/usb/device/usb_dc_stm.c]: I changed it to CONFIG_USB_DC_STM_EP_NUM.

File: drivers/usb/device/usb_dc_stm.c (hunk 1 above)
Comment (ydamigos) [reply drivers/usb/device/usb_dc_stm.c]: I updated it with the number of FIFOs.

File: drivers/usb/device/usb_dc_stm.c [hunk 6]
```diff
@@ -0,0 +1,644 @@
...
+		return -EIO;
+	}
+
+	ret = usb_dc_ep_clear_stall(ep);
+	if (ret) {
+		return ret;
+	}
+
+	if (EP_IS_OUT(ep) && ep != EP0_OUT) {
+		ret = usb_dc_ep_start_read(ep,
```
[the remark this answers is not in the document]
Comment (ydamigos) [reply drivers/usb/device/usb_dc_stm.c]: The original author tries to explain it in comments at the head of the file (lines 19 to 31).

**Comment by ydamigos** [comment]
A small update regarding the status of this PR:

I made some changes to the driver in order to support the other STM32 SOC families.
It should work with the SOCs which support OTG FS controller. I only tested it on stm32f401re.
It compiles fine for the F3, F1 (with USB device controller) but the communication fails after the SET ADDRESS setup packet.
Some boards use a USB DISCONNECT pin to enable a pullup in the USB DP signal. I will add some Kconfig options to address this issue.

**Comment by daniel-thompson** [comment]
@loicpoulain : You might want to take part in review process here.

**Comment by ydamigos** [comment]
I am closing the PR because PR #4212, which add STM32 usb device driver for F4 family, was merged and PR #4376, which adds DT support for STM32 usb device driver, is under review. I will rebase the work here against those PRs and I will create a new PR.

**Pull request closed without merging**